src/plugins/discover/public/context_awareness/profiles_manager.ts

```diff
@@ -0,0 +1,220 @@
+  private readonly recordId$ = new BehaviorSubject<string | undefined>(undefined);
+
+  private prevRootProfileParams?: SerializedRootProfileParams;
+  private prevDataSourceProfileParams?: SerializedDataSourceProfileParams;
+  private rootProfileAbortController?: AbortController;
+  private dataSourceProfileAbortController?: AbortController;
+
+  constructor(
+    private readonly rootProfileService: RootProfileService,
+    private readonly dataSourceProfileService: DataSourceProfileService,
+    private readonly documentProfileService: DocumentProfileService
+  ) {
+    this.rootContext$ = new BehaviorSubject(rootProfileService.defaultContext);
+    this.dataSourceContext$ = new BehaviorSubject(dataSourceProfileService.defaultContext);
+  }
+
+  public async resolveRootProfile(params: RootProfileProviderParams) {
+    const serializedParams = serializeRootProfileParams(params);
+
+    if (isEqual(this.prevRootProfileParams, serializedParams)) {
+      return;
+    }
+
+    const abortController = new AbortController();
+    this.rootProfileAbortController?.abort();
+    this.rootProfileAbortController = abortController;
+
+    let context = this.rootProfileService.defaultContext;
+
+    try {
+      context = await this.rootProfileService.resolve(params);
+    } catch (e) {
+      logResolutionError(ContextType.Root, serializedParams, e);
+    }
+
+    if (abortController.signal.aborted) {
+      return;
+    }
+
+    this.rootContext$.next(context);
+    this.prevRootProfileParams = serializedParams;
+  }
+
+  public async resolveDataSourceProfile(params: DataSourceProfileProviderParams) {
+    const serializedParams = serializeDataSourceProfileParams(params);
+
+    if (isEqual(this.prevDataSourceProfileParams, serializedParams)) {
+      return;
+    }
+
+    const abortController = new AbortController();
+    this.dataSourceProfileAbortController?.abort();
+    this.dataSourceProfileAbortController = abortController;
+
+    let context = this.dataSourceProfileService.defaultContext;
+
+    try {
+      context = await this.dataSourceProfileService.resolve(params);
+    } catch (e) {
+      logResolutionError(ContextType.DataSource, serializedParams, e);
+    }
+
+    if (abortController.signal.aborted) {
+      return;
+    }
+
+    this.dataSourceContext$.next(context);
+    this.prevDataSourceProfileParams = serializedParams;
+  }
+
+  public resolveDocumentProfile(params: DocumentProfileProviderParams) {
+    Object.defineProperty(params.record, 'context', {
+      get: memoize(() => {
+        let context = this.documentProfileService.defaultContext;
+
+        try {
+          context = this.documentProfileService.resolve(params);
+        } catch (e) {
+          logResolutionError(ContextType.Document, { recordId: params.record.id }, e);
+        }
+
+        return context;
+      }),
+    });
+
+    this.recordId$.next(params.record.id);
```

Comment:
Why do we collect document ids?

Reply:
Sorry, the purpose of `getProfiles$` is also not clear to me. Could you please explain the idea behind it?

Reply:
Sure, I'll have to make sure to include this when I add docs too.

The profiles manager maintains the state of all active profiles whenever one of its resolve methods is called (e.g. `resolveDataSourceProfile`). This allows retrieving the active profiles from within the React tree from a central place, and in the correct order to merge when accessing an extension point. The main place this is done is in the `use_profiles.ts` hook.

To make sure the profiles are reactive in the hook, the profiles manager exposes a `useProfiles$` observable that triggers and returns updated profiles whenever a new profile is resolved (e.g. data source profile is resolved whenever the data view changes). This way the extension points are always synced with the active profiles.

The root profile and data source profile are shared by all extension points once resolved, but the document profile depends on the document, so we pass that to the `useProfiles` hook as `record` at runtime, which gets passed down to `getProfiles$`.

Hopefully this makes things a bit more clear, but let me know if something is still unclear.

Reply:
Yes, it does. Thanks!